### ISSUE TITLE
SelectMenu placeholder capability?

### ISSUE THREAD

**Issue by jh3y** [openui/open-ui#637]
In some of the *popular* select menu offerings out there, it's often the case that you can have a default "clean" state where no value is selected: https://react-select.com/home#async

The button part tends to have some form of "placeholder" content and the options list doesn't have to contain that placeholder visually as an option.

I tried to implement this behavior with the current API: https://codepen.io/web-dot-dev/pen/qBKjYKy

The question is "How do we allow people to set a placeholder without weird hacks?"

I tried a couple of "hacks" but they break the current implementation.

- Have an option that's hidden to "trick" the button part. This works but breaks the functionality as expected. Opening the dropdown doesn't focus on an option in the dropdown.
- Using `disabled` on the first option also breaks the functionality but is a common route people take: https://stackoverflow.com/questions/5805059/how-do-i-make-a-placeholder-for-a-select-box

Do we see this as a bug in the implementation? Or should we have some way to define a placeholder? Technically allowing a SelectMenu to be in a non-selected state.

cc @mfreed7

**Comment by scottaohara** [comment]
Agree this is a common thing people want to do, and should probably be supported.  Problem with this though, and with the hacks people do to simulate it, is that it creates situations where a user cannot return to a no value selected state without resetting the form. 

If supporting a placeholder for selectmenu, it'd be nice to also account for users wanting to return to an unselected state without them having to reset the form or refresh the web page to do so.

**Comment by jh3y** [comment]
Totally agree! 💯 

Side note:: Why is there not a "💯" reaction for comments?! 😅

**Comment by scottaohara** [comment]
There are quite a number of reactions that I'd like to see as well... though, most of them fall into sarcastic/snark territory, and it's best I do not have easy access to them.

**Comment by mfreed7** [comment]
I agree that this feels like something we should handle with the new element. Are there standard ways that existing frameworks expose this functionality to developers? Just no `<option selected>` means no selection? And how is the ability to clear the selection exposed to users? It'd be good to follow cow paths for this.

**Comment by smhigley** [comment]
+1 to having a `placeholder` prop and `::placeholder` style on selectmenu. I think it'd be good to separate out clearing selection to a different issue. While `placeholder` makes the lack of clearing selection more obvious, I don't think they're otherwise related.

**Comment by josepharhar** [comment]
> masonf: agreed to move forward with some sort of placeholder, tentative to use parts and slots, any objection to move along that route?

So it sounds like we should use parts and slots to put some elements in as a placeholder and style it, but how should the selectmenu know if you want to opt into having a default unselected state with a placeholder in the first place?

> Agree this is a common thing people want to do, and should probably be supported. Problem with this though, and with the hacks people do to simulate it, is that it creates situations where a user cannot return to a no value selected state without resetting the form.

What about pressing backspace or escape? I suppose that wouldn't work on a phone when you don't have a keyboard though... perhaps an "x" button? Or an entry in the list?

**Comment by josepharhar** [comment]
The existing `<select>` element supports clearing the selected value by doing `element.value = ''` or `element.value = null`. I think this is something we could easily copy for selectmenu.

As for having a placeholder...
We currently set the selected value's textContent to an empty string when there is no selected value: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/forms/html_select_menu_element.cc;l=864-865;drc=1fa8c431e4d96c382156f7831e5246a427754705
This matches what `<select>` does, it just becomes empty when there is no selected value.
We could easily add an attribute like `placeholder` to replace that empty string with an actual string specified by the author. This attribute already exists for other elements and I think would be perfect for selectmenu: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#placeholder

If you want to do something really fancy like have the placeholder have a different style than the other selected options, I believe that falls into something that you could do with a custom placeholder element and script. It wouldn't be too different than the resolution we already made for deciding not to copy innerHTML of the selected option: https://github.com/openui/open-ui/issues/571

I propose that we support `selectmenu.value = ''` and `selectmenu.value = null` for clearing the selected value and add the `placeholder` attribute as a way of setting the textContent of the selected value element to something other than an empty string for this case.

I also propose that pressing the escape key on the button while the listbox is closed should clear the selected value because thats what aria practices says to do: https://www.w3.org/WAI/ARIA/apg/patterns/combobox/

**Comment by scottaohara** [comment]
>I also propose that pressing the escape key on the button while the listbox is closed should clear the selected value because thats what aria practices says to do: https://www.w3.org/WAI/ARIA/apg/patterns/combobox/

that's only for text-editable comboboxes though.  that'd be out of scope for selectmenu

**Comment by josepharhar** [comment]
Thanks for looking scott, I retract my proposal for the escape key behavior

**Comment by josepharhar** [comment]
At the meeting last week, people pointed out that having a placeholder attribute with a text value isn't good enough because if we find a way to display richer values in the selected value slot then simply having a text value wouldn't be as good as having all the rich content that you could put in an option element.

Scott pointed out that if you display:none or set the hidden attribute on an option, it won't show up in the dropdown, so the placeholder could simply be the first option in the list:
```html
<selectmenu>
  <option hidden>--choose an option--</option>
  <option>one</option>
  <option>two</option>
</selectmenu>
```

If you want the user to be able to go back to the unselected state, you could simply remove the hidden attribute on the first option.
Does everyone think this is a good enough solution?

**Comment by jh3y** [comment]
Hey 👋 

Does this work though? Setting `[hidden]` *kinda* works. The issue is that the value of the `selectmenu` becomes whatever that `indeterminate` option's value is. Which could be fine. But, the DX side likely expects something along the lines of `null`, `undefined`, `""`, etc.

Also, a `hidden` option [breaks the keyboard controls](https://codepen.io/jh3y/pen/oNaNqGQ/1dc404d031f4f578f0dc5b913d9a3ca2).

```
<selectmenu>
  <option value="indeterminate" hidden>Indeterminate</option>
  <option value="one" autofocus>One</option>
  <option value="two">Two</option>
  <option value="three">Three</option>
</selectmenu>
```

Overriding the `selected-value` isn't terrible depending on the design. You would check against the selectmenu value and if it's other than `null`, you update accordingly. This is purely based on the fact that `selected-value` rich content hasn't been settled on yet so assuming long term that it'll be down to the developer to manage this which isn't "terrible", jus' a little tedious.

With that said, because a `selectmenu` assumes a value of the first option, setting the value to `null` or something other than an option value breaks the browser in certain scenarios 😅 

Check [this demo](https://codepen.io/jh3y/pen/RwewMxz/3acc4c2fc01b5ff6787c4574e6a6ccfb) and tab into the control and hit up/down. Figured this was a "fun" find ha. But, this happens I believe because we've broken the tabbing controls by setting the value. 

Either of these two approaches break the keyboard tabbing. I'm all for the "indeterminate" value route as mentioned. But, the tabbing controls need to skip to the first available(visible) option 👍

**Comment by scottaohara** [comment]
The point of showing that hidden is how someone can hack this now was not to imply that it’s a perfect solution -now-, but to show that it’s a way it could be done and if it was specified to work this way, then unwanted behaviors could be mitigated

**Comment by josepharhar** [comment]
> But, the DX side likely expects something along the lines of null, undefined, "", etc.

If you set the value of the placeholder option to an empty string, then at least assigning to an empty string works: https://jsfiddle.net/jarhar/3boetfum/
Assigning to null or undefined makes it look like it has no selected value at all, which we could change to make select the first option instead.

> Check [this demo](https://codepen.io/jh3y/pen/RwewMxz/3acc4c2fc01b5ff6787c4574e6a6ccfb) and tab into the control and hit up/down

Thanks for finding this, I filed a bug to get it fixed: https://bugs.chromium.org/p/chromium/issues/detail?id=1432986

> The point of showing that hidden is how someone can hack this now was not to imply that it’s a perfect solution -now-, but to show that it’s a way it could be done and if it was specified to work this way, then unwanted behaviors could be mitigated. But if there is a better idea for how to do this, then we should go with that.

Ok, what if we put a "placeholder" attribute on the placeholder option? I forget if this was thought of during the last meeting. @jh3y do you have any ideas?

**Comment by jh3y** [comment]
> Ok, what if we put a "placeholder" attribute on the placeholder option? I forget if this was thought of during the last meeting. @jh3y do you have any ideas?

To be honest, if I was going with an explicit placeholder that isn't an option, then perhaps it becomes another optional piece associated with a `behavior`. I gather the `slot`, `behavior`, etc. part was being discussed again. But, it could make sense as its own designated element. That's how I've gone about implementing it in some demos where I've needed it. That way you could have some logic such as if a placeholder is present, don't set a default value of the `:selected` or first option. And once there is a selected value, the placeholder loses visibility or something. In some cases to get around this, I have to manually create a `selected-value` element and control it myself.

**Comment by una** [comment]
I chatted with @mfreed7 and @josepharhar about a potential placeholder solution which is relatively simple and solves a lot of open questions. 

**TLDR; Align with what `select` does for placeholders: put `value=""` on an option to indicate a placeholder**

Here's what this might look like:

Text-only placeholder:

```html
<selectlist>
  <option value="">
    Select a country
  </option>
  <option value="south-africa">
    <img src="s-a-flag.png" />
    <p>South Africa</p>
  </option>
  <option>
    ...
  </option>
</selectlist>
```

Placeholder with an image:

```html
<selectlist>
  <option value="">
    <img src="placeholder-flag.png" />
    <p>Select a country</p>
  </option>
  <option value="south-africa">
    <img src="s-a-flag.png" />
    <p>South Africa</p>
  </option>
  <option>
    ...
  </option>
</selectlist>
```

**Comment by brechtDR** [comment]
Interesting idea @una 

Few questions come to mind

1) does this mean that the empty valued option should also be the first option?

2) what if multiple empty values are provided?

3) should we consider the "default" attribute for this as well? (Generally speaking)

Would love to hear your thoughts.

**Comment by una** [comment]
1. Yes, it should. Currently, select will show "Disco" as the placeholder here:

```
<select> 
  <option value="disco">Disco</option>
  <option value="">Select a value</option>
</select>
```

2. Currently, select will choose the first option (if there isn't a `selected` attribute on it). I.e. the placeholder in the below example is "Select a value"

```
<select>
  <option value="">Select a value</option>
  <option value="">Select a value 2</option>
  <option value="disco">Disco</option>
</select>
```

3. Yes, that would be the ideal, just like with select:

```
<select> 
  <option value="" selected disabled>Select a value</option>
  <option value="disco">Disco</option>
</select>
```

**Comment by brechtDR** [comment]
@una sounds about right. 👍 

It's consistent with the current select implementation. I do remember that in the past a question was raised in a telecon with "can we do better?" 

The use case for that is that people might want to use a placeholder attribute on a select as well just as they can do with an input element. 

However, this use case doesn't have the ability to use an image as in your proposal. Especially for custom styling, I'm more into what you are proposing.

It does make me wonder: should we also allow a placeholder attribute? And what would happen if we have a placeholder combined with the empty value solution, but maybe this is beside the question.

**Comment by una** [comment]
@brechtDR I recall we've talked about a `placeholder` attribute before but it didn't really get anywhere. I wouldn't be against it, but would need to define its utility and if its worth adding a new attribute to HTML, plus how it would interact with the existing `select` options, etc. This solution also doesn't disallow `placeholder` in the future, it could work with it. i.e.


```
<selectlist>
  <option value="" placeholder>
    <img src="placeholder-flag.png" />
    <p>Select a country</p>
  </option>
  <option value="south-africa">
    <img src="s-a-flag.png" />
    <p>South Africa</p>
  </option>
  <option>
    ...
  </option>
</selectlist>
```

where `placeholder` would replace `selected disabled`

**Comment by josepharhar** [comment]
Considering that option elements already support `selected` and `disabled`, I think that we can just keep using those instead of making `placeholder` replace them. Besides, what if someone wants to make a placeholder than the user can go back to and select again? In this case, they could use `selected` without `disabled`.

I'd like to propose these three behaviors, aligning with @una, all three of which are already match the existing `<select>` element:
1. `<option value="">` in a `<selectlist required>` prevents form submission when selected.
2. `<option disabled>` prevents the option from being selected by the user. It can still be the default initially selected option.
3. `<option selected>` makes the option the default initially selected option. If there is no `<option selected>`, then the first child option in the DOM is the default initially selected option. If there are multiple `<option selected>`s, then the first one in the DOM wins.

Given all these behaviors, I think that authors should be able to do any placeholder use case they want.

**Comment by brechtDR** [comment]
> imo `<option hidden>` should be fine. 

In combination with this, it sounds really powerful. Absolutely great idea!

**Comment by lukewarlow** [comment]
For the placeholder aspect (wouldn't cover the reset one).

Could we allow `<selectedoption>Placeholder content goes in here<selectedoption>` kinda like how a slot takes default content?

**Comment by josepharhar** [comment]
> Could we allow `<selectedoption>Placeholder content goes in here<selectedoption>` kinda like how a slot takes default content?

If we end up cloning into the user-agent shadowroot of the selectedoption element then I suppose that's a possibility. However, the way its currently prototyped and the way that some of us are pushing for is to replace the light-dom contents of the selectedoption element when cloning the selected option element, meaning that the "Placeholder content goes ..." text child of the selectedoption element will always get removed by the browser.

**Comment by josepharhar** [comment]
Is all we need to discuss here the idea of putting content inside the selectedoption element? I still think it would be better to follow the paradigm that the existing select element has with its options to do placeholders, especially now that we are re-using the select element.
